src/java/main/org/apache/zookeeper/server/ContainerManager.java

```diff
@@ -0,0 +1,133 @@
+ * be run from the leader. There's no harm in running from followers/observers
+ * but that will be extra work that's not needed. Once started, it periodically
+ * checks container nodes that have a cversion > 0 and have no children. A delete
+ * is attempted on the node. The result of the delete is unimportant. If the proposal
+ * fails or the container node is not empty there's no harm.
+ */
+public class ContainerManager {
+    private static final Logger LOG = LoggerFactory.getLogger(ContainerManager.class);
+    private final ZKDatabase zkDb;
+    private final RequestProcessor requestProcessor;
+    private final int checkIntervalMs;
+    private final int maxPerInterval;
+    private final Timer timer;
+    private final AtomicReference<TimerTask> task = new AtomicReference<TimerTask>(null);
+    private final Set<String> currentCandidates = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+
+    /**
+     * @param zkDb the ZK database
+     * @param requestProcessor request processer - used to inject delete container requests
+     * @param checkIntervalMs how often to check containers in milliseconds
+     * @param maxPerInterval the max containers to delete in one interval - avoids herding of container deletions
+     */
+    public ContainerManager(ZKDatabase zkDb, RequestProcessor requestProcessor,
+                            int checkIntervalMs, int maxPerInterval) {
+        this.zkDb = zkDb;
+        this.requestProcessor = requestProcessor;
+        this.checkIntervalMs = checkIntervalMs;
+        this.maxPerInterval = maxPerInterval;
+        timer = new Timer("ContainerManagerTask", true);
+    }
+
+    /**
+     * start/restart the timer the runs the check. Can safely be called multiple times.
+     */
+    public void start() {
+        if ( task.get() == null ) {
+            TimerTask timerTask = new TimerTask() {
+                @Override
+                public void run() {
+                    checkContainers();
+                }
+            };
+            if ( task.compareAndSet(null, timerTask) ) {
+                timer.scheduleAtFixedRate(timerTask, checkIntervalMs, checkIntervalMs);
+            }
+        }
+    }
+
+    /**
+     * stop the timer if necessary. Can safely be called multiple times.
+     */
+    public void stop() {
+        TimerTask timerTask = task.getAndSet(null);
+        if ( timerTask != null ) {
+            timerTask.cancel();
+        }
+    }
+
+    /**
+     * Manually check the containers. Not normally used directly
+     */
+    public void checkContainers() {
+        if ( currentCandidates.size() == 0 ) {
+            currentCandidates.addAll(getCandidates());
+        }
+
+        int count = 0;
+        Iterator<String> iterator = currentCandidates.iterator();
+        while ( iterator.hasNext() ) {
+            if ( count++ >= maxPerInterval ) {
+                LOG.info("Stopping at " + maxPerInterval);
```

Comment:
I'm thinking of changing maxPerInterval to maxPerSecond. Then, checkContainers() would process all candidates by sleep between each deletion so that there's a capped TPS. Thoughts?